presto-main/src/main/java/com/facebook/presto/operator/PagesSpatialIndexSupplier.java

```diff
@@ -0,0 +1,152 @@
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(PagesSpatialIndexSupplier.class).instanceSize();
+    private static final int ENVELOPE_INSTANCE_SIZE = ClassLayout.parseClass(Envelope.class).instanceSize();
+    private static final int STRTREE_INSTANCE_SIZE = ClassLayout.parseClass(STRtree.class).instanceSize();
+    private static final int ABSTRACT_NODE_INSTANCE_SIZE = ClassLayout.parseClass(AbstractNode.class).instanceSize();
+
+    private final Session session;
+    private final LongArrayList addresses;
+    private final List<Type> types;
+    private final List<Integer> outputChannels;
+    private final List<List<Block>> channels;
+    private final BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest;
+    private final Optional<JoinFilterFunctionCompiler.JoinFilterFunctionFactory> filterFunctionFactory;
+    private final STRtree rtree;
+    private final long memorySizeInBytes;
+
+    public PagesSpatialIndexSupplier(
+            Session session,
+            LongArrayList addresses,
+            List<Type> types,
+            List<Integer> outputChannels,
+            List<List<Block>> channels,
+            int geometryChannel,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest,
+            Optional<JoinFilterFunctionCompiler.JoinFilterFunctionFactory> filterFunctionFactory)
+    {
+        this.session = session;
+        this.addresses = addresses;
+        this.types = types;
+        this.outputChannels = outputChannels;
+        this.channels = channels;
+        this.spatialRelationshipTest = spatialRelationshipTest;
+        this.filterFunctionFactory = filterFunctionFactory;
+
+        this.rtree = buildRTree(addresses, channels, geometryChannel);
+        this.memorySizeInBytes = INSTANCE_SIZE +
+                (rtree.isEmpty() ? 0 : STRTREE_INSTANCE_SIZE + computeMemorySizeInBytes(rtree.getRoot())) +
+                sizeOf(addresses.elements()) +
+                channels.stream()
+                    .flatMap(List::stream)
+                    .mapToLong(Block::getRetainedSizeInBytes)
+                    .sum();
+    }
+
+    private static STRtree buildRTree(LongArrayList addresses, List<List<Block>> channels, int geometryChannel)
+    {
+        STRtree rtree = new STRtree();
+
+        for (int position = 0; position < addresses.size(); position++) {
+            long pageAddress = addresses.getLong(position);
+            int blockIndex = decodeSliceIndex(pageAddress);
+            int blockPosition = decodePosition(pageAddress);
+
+            Block block = channels.get(geometryChannel).get(blockIndex);
+            // TODO Consider pushing is-null and is-empty checks into a filter below the join
+            if (block.isNull(blockPosition)) {
+                continue;
+            }
+
+            Slice slice = block.getSlice(blockPosition, 0, block.getSliceLength(blockPosition));
+            OGCGeometry ogcGeometry = deserialize(slice);
```

Comment:
can `ogcGeometry` be null (deserialize can return null)?

Reply:
`deserialize` returns `null` only if input is `null`.  Here we know that input is not `null` because of the check above.